### PR TITLE
Addition of Serial node

### PR DESCRIPTION
**Pull request by dstrande**
### Description

SERIAL is a node allowing for pulling data off a serial device like a Arduino. 

SELECT_ARRAY is a node for choosing which column of a numpy array to use for y data.

SELECT_ARRAY must be used after SERIAL (even if only column is used). 

### Inputs & Outputs

SERIAL - no inputs, list output for x, numpy array output for y. 

### Example APPS

serial.txt

### Integrating the Node

It should contain checklist of steps to successfully integrate your node into the framework.

### CODE REVIEW

File: INSTRUMENTS/Serial/Serial/SERIAL.py [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+            if len(times) > 0:
+                time1 = seconds - times[i]
+            else:
+                time1 = 0.1  # Estimate of execution time 
+
+            if time1 < RECORD_PERIOD:
+                # print(f'sleeping for {RECORD_PERIOD - time1} seconds.')
+                sleep(RECORD_PERIOD - time1)
+
+    # print('')
```
Comment (smahmed776) [inline INSTRUMENTS/Serial/Serial/SERIAL.py:66]: I think we should remove all unnecessary print statements.

**Review by smahmed776 — COMMENTED** [review]
Looks good, However, it is recommended to remove any unnecessary print statements to enhance its cleanliness.

**Comment by jackparmer** [comment]
@Angchau can you please try to replicate this? There's an Arduino in the office.

File: INSTRUMENTS/Serial/Serial/SERIAL.py [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+def SERIAL(v, params):
+    '''
+    Node to take simple time dependent 1d data from an Ardunio, or a similar serial device.
+    For example you can record temperature following this tutorial:
+
+    https://learn.adafruit.com/thermistor/using-a-thermistor
+
+    with Serial.println(steinhart) as the only line printing.
+
+    It is important that the last line Arduino is returning is the data with a new line at the end (i.e. println()).
```
Comment (jackparmer) [inline INSTRUMENTS/Serial/Serial/SERIAL.py:18]: exceeds the PEP-8 max line length
https://stackoverflow.com/questions/88942/why-does-pep-8-specify-a-maximum-line-length-of-79-characters

please install a PEP-8 linter

File: INSTRUMENTS/Serial/Serial/SERIAL.py [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+    print(reading0)
+    print(",")
+    println(reading1)
+    '''
+    print('parameters passed to SERIAL: ', params)
+    COM_PORT = params.get('com_port', '/dev/ttyUSB0') # The COM port to use (e.g.: Linux/Mac: '/dev/ttyUSB0' or Windows: COM1).
+    BAUD = int(params.get('BAUD_RATE', 9600))
+    NUM = int(params.get('num_readings', 100)) # Number of points to record. Length of recording ~= NUM * RECORD_PERIOD seconds.
+    RECORD_PERIOD = float(params.get('record_period', 1)) # Period of recordings in seconds (i.e. the time between two recordings).
+    # NUM = 5  # Debug
```
Comment (jackparmer) [inline INSTRUMENTS/Serial/Serial/SERIAL.py:33]: delete uneccesary comments

File: TRANSFORMERS/SELECT_ARRAY/SELECT_ARRAY/SELECT_ARRAY.py [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+
+@flojoy
+def SELECT_ARRAY(v, params):
+    '''
+    Node to convert an input array with multiple columns to the selected ordered pair.
+
+    For example, the SERIAL node can output x=time, y1=temperature, y2=pressure.
+    This node will select one of temperautre and pressure columns to output.
+    '''
+    print('parameters passed to SELECT_ARRAY: ', params)
```
Comment (jackparmer) [inline TRANSFORMERS/SELECT_ARRAY/SELECT_ARRAY/SELECT_ARRAY.py:11]: delete

File: TRANSFORMERS/SELECT_ARRAY/SELECT_ARRAY/SELECT_ARRAY.py [hunk 5]
```diff
@@ -0,0 +1,35 @@
+from flojoy import flojoy, DataContainer
+
+@flojoy
+def SELECT_ARRAY(v, params):
+    '''
+    Node to convert an input array with multiple columns to the selected ordered pair.
+
+    For example, the SERIAL node can output x=time, y1=temperature, y2=pressure.
+    This node will select one of temperautre and pressure columns to output.
```
Comment (jackparmer) [inline TRANSFORMERS/SELECT_ARRAY/SELECT_ARRAY/SELECT_ARRAY.py:9]: temperature

**Review by jackparmer — CHANGES_REQUESTED** [review]
please remove all of the comments that were for debugging

also please make sure adheres to PEP-8 for readability and cleanlieness
https://gist.github.com/RichardBronosky/454964087739a449da04

File: INSTRUMENTS/Serial/Serial/SERIAL.py (hunk 1 above)
Comment (dstrande) [reply INSTRUMENTS/Serial/Serial/SERIAL.py]: Debugging comments and print statements removed.

File: INSTRUMENTS/Serial/Serial/SERIAL.py (hunk 2 above)
Comment (dstrande) [reply INSTRUMENTS/Serial/Serial/SERIAL.py]: PEP-8 changes made.

File: INSTRUMENTS/Serial/Serial/SERIAL.py (hunk 3 above)
Comment (dstrande) [reply INSTRUMENTS/Serial/Serial/SERIAL.py]: Debugging comments and print statements removed.

File: TRANSFORMERS/SELECT_ARRAY/SELECT_ARRAY/SELECT_ARRAY.py (hunk 4 above)
Comment (dstrande) [reply TRANSFORMERS/SELECT_ARRAY/SELECT_ARRAY/SELECT_ARRAY.py:17]: Debugging comments and print statements removed.

File: TRANSFORMERS/SELECT_ARRAY/SELECT_ARRAY/SELECT_ARRAY.py (hunk 5 above)
Comment (dstrande) [reply TRANSFORMERS/SELECT_ARRAY/SELECT_ARRAY/SELECT_ARRAY.py]: Done. Is there an extension or linter to check for typos in docstrings?

**Comment by dstrande** [comment]
Trying to figure out what failed here. 

"Style / style (pull_request)" Is this PEP-8 or something else?

Strange it didn't fail for my first commit.

**Pull request merged**